devito/types/tensor.py

```diff
@@ -0,0 +1,598 @@
+    diagonal : Bool, optional
+        Wether the tensor is diagonal or not. Defaults to False.
+    staggered: Tuple of Dimension, optional
+        Staggering of each component, needs to have the size of the tensor. Defaults
+        to the Dimensions.
+    """
+    _sub_type = Function
+    _op_priority = Differentiable._op_priority + 2.
+    _class_priority = 10
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            self._is_symmetric = kwargs.get('symmetric', True)
+            self._is_diagonal = kwargs.get('diagonal', False)
+            self._staggered = kwargs.get('staggered', self.space_dimensions)
+            self._grid = kwargs.get('grid')
+            self._space_order = kwargs.get('space_order', 1)
+
+    @classmethod
+    def __setup_subfunc__(cls, *args, **kwargs):
+        """
+        Creates the components of the TensorFunction
+        either from input or from input Dimensions.
+        """
+        comps = kwargs.get("components")
+        if comps is not None:
+            return comps
+        funcs = []
+        grid = kwargs.get("grid")
+        if grid is None:
+            dims = kwargs.get('dimensions')
+            if dims is None:
+                raise TypeError("Need either `grid` or `dimensions`")
+        else:
+            dims = grid.dimensions
+        stagg = kwargs.get("staggered", None)
+        name = kwargs.get("name")
+        symm = kwargs.get('symmetric', True)
+        # Fill tensor, only upper diagonal if symmetric
+        for i, d in enumerate(dims):
+            start = i+1 if symm else 0
+            funcs2 = [0 for _ in range(i+1)] if symm else []
+            for j in range(start, len(dims)):
+                kwargs["name"] = "%s_%s%s" % (name, d.name, dims[j].name)
+                kwargs["staggered"] = (stagg[i][j] if stagg is not None
+                                       else (NODE if i == j else (d, dims[j])))
+                funcs2.append(cls._sub_type(**kwargs))
+            funcs.append(funcs2)
+
+        # Symmetrize and fill diagonal if symmetric
+        if symm:
+            funcs = np.array(funcs) + np.array(funcs).T
+            for i in range(len(dims)):
+                kwargs["name"] = "%s_%s%s" % (name, dims[i].name, dims[i].name)
+                kwargs["staggered"] = stagg[i][i] if stagg is not None else NODE
+                funcs[i, i] = cls._sub_type(**kwargs)
+            funcs = funcs.tolist()
+        return funcs
+
+    def __getattr__(self, name):
+        """
+        Try calling a dynamically created FD shortcut.
+
+        Notes
+        -----
+        This method acts as a fallback for __getattribute__
+        """
+        if name in self[0]._fd:
+            return self.applyfunc(lambda x: getattr(x, name))
+        raise AttributeError("%r object has no attribute %r" % (self.__class__, name))
+
+    @call_highest_priority('__rmul__')
+    def __mul__(self, other):
```

Comment:
before I review this: this is completely new code right? it's not something taken from SymPy and slightly adapted?

Reply:
Yep